kernel/src/net/socket/netlink/kobject_uevent/message/syn_uevent.rs

```diff
@@ -0,0 +1,112 @@
+//! It differs from the event triggers by devices.
+//!
+//! Reference: <https://elixir.bootlin.com/linux/v6.0.9/source/Documentation/ABI/testing/sysfs-uevent>.
+
+use alloc::format;
+use core::str::FromStr;
+
+use super::uevent::SysObjAction;
+use crate::prelude::*;
+
+/// The synthetic uevent.
+pub struct SyntheticUevent {
+    pub(super) action: SysObjAction,
+    pub(super) uuid: Option<Uuid>,
+    pub(super) envs: BTreeMap<String, String>,
+}
+
+impl FromStr for SyntheticUevent {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        let mut split = s.split(" ");
+
+        let action = {
+            let Some(action_str) = split.next() else {
+                return_errno_with_message!(Errno::EINVAL, "the string is empty");
+            };
+            SysObjAction::from_str(action_str)?
+        };
+
+        let uuid = if let Some(uuid_str) = split.next() {
+            Some(Uuid::from_str(uuid_str)?)
+        } else {
+            None
+        };
+
+        let mut envs = BTreeMap::new();
+        for env_str in split.into_iter() {
+            let (key, value) = {
+                // Each string should be in the `KEY=VALUE` format.
+                let mut split = env_str.split('=');
+
+                let (key, value) = (split.next(), split.next());
+                if key.is_none() || value.is_none() || split.next().is_some() {
+                    return_errno_with_message!(Errno::EINVAL, "invalid key value pairs");
+                }
+                (key.unwrap(), value.unwrap())
+            };
+
+            // Both `KEY` and `VALUE` can contain alphanumeric characters only.
+            for byte in key.as_bytes().iter().chain(value.as_bytes()) {
+                if !byte.is_ascii_alphanumeric() {
+                    return_errno_with_message!(
+                        Errno::EINVAL,
+                        "invalid character in key value pairs"
+                    );
+                }
+            }
+
+            // The `KEY` name gains ``SYNTH_ARG_`` prefix to avoid possible collisions
+            // with existing variables.
+            let key = format!("SYNTH_ARG_{}", key);
+            let value = value.to_string();
+            envs.insert(key, value);
+        }
+
+        Ok(Self { action, uuid, envs })
+    }
+}
+
+pub struct Uuid(pub(super) String);
+
+impl FromStr for Uuid {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        // The allowed UUID pattern is:
+        // xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+        // where each `x` is a hex digit.
+        // Reference: <https://elixir.bootlin.com/linux/v6.0.9/source/Documentation/ABI/testing/sysfs-uevent#L19>.
+
+        const UUID_LEN: usize = 36; // 8+1+4+1+4+1+4+1+12
+
+        let bytes = s.as_bytes();
+
+        if bytes.len() != UUID_LEN {
+            return_errno_with_message!(Errno::EINVAL, "the UUID length is invalid");
+        }
+
+        const HYPHEN_INDEXES: [usize; 4] = [8, 13, 18, 23];
```

Comment:
Something like this may be better:
```rust
    /// The allowed UUID pattern, where each `x` is a hex digit.
    ///
    /// Reference: <https://elixir.bootlin.com/linux/v6.0.9/source/Documentation/ABI/testing/sysfs-uevent#L19>.
    const UUID_PATTERN: &str = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";

    if bytes.len() != UUID_PATTERN.len() {
         return_errno_with_message!(Errno::EINVAL, "the UUID length is invalid");
    }

    for (byte, pattern) in bytes.iter().zip(UUID_PATTERN.iter()) {
        if *byte == 'x' && pattern.is_ascii_hexdigit() {
            continue;
        } else if *byte == '-' && pattern == '-' {
            continue;
        } else {
            return_errno_with_message!(Errno::EINVAL, "the UUID content is invalid");
        }
    }
```